llama/xla_model_parallel.py

```diff
@@ -350,6 +350,105 @@ def forward(self, input_: torch.Tensor) -> torch.Tensor:  # type: ignore
+                       which is Y_i = XA_i
+        init_method: method to initialize weights. Note that bias is always set
+                     to zero.
+        stride: For the strided linear layers.
+        keep_master_weight_for_test: This was added for testing and should be
+                                     set to False. It returns the master weights
+                                     used for initialization.
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        bias: bool = True,
+        gather_output: bool = True,
+        init_method: Callable[[torch.Tensor], torch.Tensor] = init.xavier_normal_,
+        stride: int = 1,
+        keep_master_weight_for_test: bool = False,
+        world_size: Optional[int] = None,
+        rank: Optional[int] = None,
+        groups: Optional[List] = None,
+    ) -> None:
+        super(ColumnParallelLinearQuant, self).__init__()
+
+        if world_size is None:
+            self.groups = get_model_parallel_group()
+            self.world_size = get_model_parallel_world_size()
+            self.rank = get_model_parallel_rank()
+        else:
+            self.groups = groups
+            self.world_size = world_size
+            self.rank = rank
+
+        # Keep input parameters
+        self.in_features = in_features
+        self.out_features = out_features
+        self.gather_output = gather_output
+        # Divide the weight matrix along the last dimension.
+        world_size = get_model_parallel_world_size()
+        self.output_size_per_partition = divide_and_check_no_remainder(out_features, self.world_size)
+
+        # Parameters.
+        # Note: torch.nn.functional.linear performs XA^T + b and as a result
+        # we allocate the transpose.
+        # self.weight = Parameter(torch.Tensor(self.output_size_per_partition, self.in_features))
+        self.weight = Parameter(torch.empty((self.output_size_per_partition, self.in_features), dtype=torch.int8), requires_grad=False)
+        self.weight_scaler = Parameter(torch.empty(1), requires_grad=False)
+        if bias:
+            self.bias = Parameter(torch.Tensor(self.output_size_per_partition))
+            # Always initialize bias to zero.
+            with torch.no_grad():
+                self.bias.zero_()
+        else:
+            self.register_parameter("bias", None)
+
+        # Initialize weight.
+        self.master_weight = _initialize_affine_weight(
+            self.weight,
+            self.out_features,
+            self.in_features,
+            self.output_size_per_partition,
+            0,
+            init_method,
+            self.world_size,
+            self.rank,
+            stride=stride,
+            return_master_weight=keep_master_weight_for_test,
+        )
+        self.groups = get_model_parallel_group()
+        self.world_size = get_model_parallel_world_size()
+        self.rank = get_model_parallel_rank()
```

Comment:
These lines should have been deleted in the non-quantized versions

Reply:
Missed during rebase, let me update. Thanks for catching it!